File: Javascript/Heap.js
class Heap{
    /* Heap constructor */
    constructor(){
        this.heap = [];
    }

    // return a specified node parent node
    parent(i){
        if (i == 0){
            return 0;
        }
        return Math.floor((i - 1) / 2);
    }

    // return a specified node left node
    left(i){
        return (i * 2) + 1;
    }

    // return a specified node right node
    right(i){
        return (i * 2) + 1;
    }

    /* heappush(x)
     * push the value [x] onto the heap, maintaining the heap invariant
     */
    heappush(x){
        // if heap is empty
        if (this.heap.length == 0){
            this.heap.push(x);
            return;
        }

        this.heap.push(x);
        var cur = this.heap.length - 1;
        var par = this.parent(cur);
        while (this.heap[cur] > this.heap[par]){
            var tmp = this.heap[cur];
            this.heap[cur] = this.heap[par];
            this.heap[par] = tmp;
            cur = par;
            par = this.parent(cur);
        }
    }

    /* heappop()
     * pop and return the greatest item from the heap
     */
    heappop(){
        // if heap is empty
        if (this.heap.length == 0){
            return;
        }
        // if heap has only one item
        if (this.heap.length == 1){
            var data = this.heap[0];
            this.heap = [];
            return data;
        }

        var data = this.heap[0];
        var tmp = this.heap[0];
        this.heap[0] = this.heap[this.heap.length - 1];
        this.heap[this.heap.lenght - 1] = tmp;
        this.heap = this.heap.slice(0, -1);
        var cur = 0;
        while (cur < this.heap.length){
            var l = this.left(cur);
            var r = this.right(cur);
            if ((l > this.heap.length - 1) || (r > this.heap.length - 1)){
                break;
            }
            if ((this.heap[cur] > this.heap[l] && this.heap[cur] > this.heap[r])){
                break;
            }
            if ((this.heap[cur] < this.heap[l] && this.heap[cur] < this.heap[r])){
                if (this.heap[l] > this.heap[r]){
                    var a = this.heap[l];
                    this.heap[l] = this.heap[cur];
                    this.heap[cur] = a;
                    cur = l;
                }
                else{
                    var b = this.heap[r];
                    this.heap[r] = this.heap[cur];
                    this.heap[cur] = b;
                    cur = r;
                }
            }
            else if (this.heap[cur] < this.heap[l]){
                var c = this.heap[l];
                this.heap[l] = this.heap[cur];
                this.heap[cur] = c;
                cur = l;
            }
            else{
                var d = this.heap[r];
                this.heap[r] = this.heap[cur];
                this.heap[cur] = d;
                cur = r;
            }
        }
        return data;
    }

    /* heapreplace(x)
     * pop and return max item
     * push [x] on the heap, efficiently
     */
    heapreplace(x){
        // if heap is empty
        if (this.heap.length == 0){
            return;
        }
        // if heap has only one item
        if (this.heap.length == 1){
            var data = this.heap[0];
            this.heap = [];
            return data;
        }

        var data = this.heap[0];
        this.heap[0] = x; 
        var cur = 0;
        while (cur < this.heap.length){
            var l = this.left(cur);
            var r = this.right(cur);
            if ((l > this.heap.length - 1) || (r > this.heap.length - 1)){
                break;
            }
            if ((this.heap[cur] > this.heap[l] && this.heap[cur] > this.heap[r])){
                break;
            }
            if ((this.heap[cur] < this.heap[l] && this.heap[cur] < this.heap[r])){
                if (this.heap[l] > this.heap[r]){
                    var a = this.heap[l];
                    this.heap[l] = this.heap[cur];
                    this.heap[cur] = a;
                    cur = l;
                }
                else{
                    var b = this.heap[r];
                    this.heap[r] = this.heap[cur];
                    this.heap[cur] = b;
                    cur = r;
                }
            }
            else if (this.heap[cur] < this.heap[l]){
                var c = this.heap[l];
                this.heap[l] = this.heap[cur];
                this.heap[cur] = c;
                cur = l;
            }
            else{
                var d = this.heap[r];
                this.heap[r] = this.heap[cur];
                this.heap[cur] = d;
                cur = r;
            }
        }
        return data;
    }

    /* print current heap as a list */
    print(){
        if (this.heap.length == 0){
            console.log("Empty Heap");
        }
        process.stdout.write("Root -> ");
        console.log(this.heap);
    }
}

var hp = new Heap();
process.stdout.write("Current Heap: ");
hp.print();

console.log("Heappush -> 39");
console.log("Heappush -> 100");
console.log("Heappush -> 78");
console.log("Heappush -> 70");
console.log("Heappush -> 33");
console.log("Heappush -> 35");
console.log("Heappush -> 54");
console.log("Heappush -> 13");
console.log("Heappush -> 59");
hp.heappush(39);
hp.heappush(100);
hp.heappush(78);
hp.heappush(70);
hp.heappush(33);
hp.heappush(35);
hp.heappush(54);
hp.heappush(13);
hp.heappush(59);
process.stdout.write("Current Heap: ");
hp.print();

console.log("Heappop ->");
process.stdout.write("The greatest item is ");
console.log(hp.heappop());
process.stdout.write("Current Heap: ");
hp.print();

console.log("Heapreplace -> 43");
process.stdout.write("The greatest item is ");
console.log(hp.heapreplace());
process.stdout.write("Current Heap: ");
hp.print();

File: Python/Heap.py
class Heap:
    # Heap constructor
    def __init__(self):
        self.heap = []
    
    # return a specified node parent node
    def parent(self, i):
        if i == 0:
            return 0
        return (i - 1) // 2
    
    # return a specified node left node
    def left(self, i):
        return (2 * i) + 1
    
    # return a specified node right node
    def right(self, i):
        return (2 * i) + 2
    
    # heappush(x)
    # push the value [x] onto the heap, maintaining the heap invariant
    def heappush(self, x):
        # if heap is empty
        if len(self.heap) == 0:
            self.heap.append(x)
            return
        
        self.heap.append(x)
        cur = len(self.heap) - 1
        par = self.parent(cur)
        while self.heap[cur] > self.heap[par]:
            self.heap[cur], self.heap[par] = self.heap[par], self.heap[cur]
            cur = par
            par = self.parent(cur)
    
    # heappop()
    # pop and return the greatest item from the heap
    def heappop(self):
        # if heap is empty
        if len(self.heap) == 0:
            return
        # if heap has only one item
        if len(self.heap) == 1:
            data = self.heap[0]
            self.heap = []
            return data
        
        self.heap[0], self.heap[-1] = self.heap[-1], self.heap[0]
        data = self.heap[-1]
        self.heap = self.heap[:-1]
        cur = int(0)
        while cur < len(self.heap):
            l = self.left(cur)
            r = self.right(cur)
            if l > len(self.heap) - 1 or r > len(self.heap) - 1:
                break
            if self.heap[cur] > self.heap[l] and self.heap[cur] > self.heap[r]:
                break

            if self.heap[cur] < self.heap[l] and self.heap[cur] < self.heap[r]:
                if self.heap[l] > self.heap[r]:
                    self.heap[cur], self.heap[l] = self.heap[l], self.heap[cur]
                    cur = l
                else:
                    self.heap[cur], self.heap[r] = self.heap[r], self.heap[cur]
                    cur = r
            elif self.heap[cur] < self.heap[l]:
                self.heap[cur], self.heap[l] = self.heap[l], self.heap[cur]
                cur = l
            else:
                self.heap[cur], self.heap[r] = self.heap[r], self.heap[cur]
                cur = r
        return data
    
    # heapreplace(x)
    # pop and return max item
    # push [x] on the heap, efficiently
    def heapreplace(self, x):
        # if heap is empty
        if len(self.heap) == 0:
            return
        # if heap has only one item
        if len(self.heap) == 1:
            data = self.heap[0]
            self.heap = []
            return data
        
        data = self.heap[0]
        self.heap[0] = x
        cur = int(0)
        while cur < len(self.heap):
            l = self.left(cur)
            r = self.right(cur)
            if l > len(self.heap) - 1 or r > len(self.heap) - 1:
                break
            if self.heap[cur] > self.heap[l] and self.heap[cur] > self.heap[r]:
                break

            if self.heap[cur] < self.heap[l] and self.heap[cur] < self.heap[r]:
                if self.heap[l] > self.heap[r]:
                    self.heap[cur], self.heap[l] = self.heap[l], self.heap[cur]
                    cur = l
                else:
                    self.heap[cur], self.heap[r] = self.heap[r], self.heap[cur]
                    cur = r
            elif self.heap[cur] < self.heap[l]:
                self.heap[cur], self.heap[l] = self.heap[l], self.heap[cur]
                cur = l
            else:
                self.heap[cur], self.heap[r] = self.heap[r], self.heap[cur]
                cur = r
        return data


    # print current heap as a list
    def print(self):
        if len(self.heap) == 0:
            print("Empty Heap")
            return
        
        print("Root -> ", end = "")
        print(self.heap)


hp = Heap()
print("Current Heap: ", end = "")
hp.print()

print("Heappush -> 39")
print("Heappush -> 100")
print("Heappush -> 78")
print("Heappush -> 70")
print("Heappush -> 33")
print("Heappush -> 35")
print("Heappush -> 54")
print("Heappush -> 13")
print("Heappush -> 59")
hp.heappush(39)
hp.heappush(100)
hp.heappush(78)
hp.heappush(70)
hp.heappush(33)
hp.heappush(35)
hp.heappush(54)
hp.heappush(13)
hp.heappush(59)
print("Current Heap: ", end = "")
hp.print()

print("Heappop ->")
print("The greatest item is ", end = "")
print(hp.heappop())
print("Current Heap: ", end = "")
hp.print()

print("Heapreplace -> 43")
print("The greatest item is ", end = "")
print(hp.heapreplace(43))
print("Current Heap: ", end = "")
hp.print()
